computedfields/resolver.py

```diff
@@ -0,0 +1,806 @@
+            to detect cyclic dependencies and might raise a ``CycleNodeException``.
+
+            If you experience this in your project try to get in-depth cycle
+            information, either by using the ``rendergraph`` management command or
+            by directly accessing the graph objects:
+
+            - intermodel dependency graph: ``your_model._graph``
+            - mode local dependency graphs: ``your_model._graph.modelgraphs[your_model]``
+            - union graph: ``your_model._graph.get_uniongraph()``
+
+            Also see the graph documentation :ref:`here<graph>`.
+        """
+        def wrap(func):
+            field._computed = {
+                'func': func,
+                'depends': depends or [],
+                'select_related': select_related,
+                'prefetch_related': prefetch_related
+            }
+            field.editable = False
+            self.add_field(field)
+            return field
+        return wrap
+
+    def update_computedfields(self, instance, update_fields=None):
+        """
+        Update values of local computed fields of ``instance``.
+
+        Other than calling ``.compute`` on an instance, this call overwrites
+        computed field values.
+
+        Returns ``None`` or an updated set of field names for ``update_fields``.
+
+        .. NOTE::
+
+            Normally you dont need to call this method yourself, if you respect
+            the inheritance remarks in :class:`.models.ComputedFieldsModel`.
+            Still for complicated late field access and changes in a custom `save` method
+            it might be necessary. For this consult the source code of
+            :meth:`save<.models.ComputedFieldsModel.save>` and
+            :meth:`precomputed<.resolver.Resolver.precomputed>` to get an idea,
+            how to correctly handle `update_fields`.
+        """
+        model = type(instance)
+        if not self.has_computedfields(model):
+            return update_fields
+        cf_mro = self.get_local_mro(model, update_fields)
+        if update_fields:
+            update_fields = set(update_fields)
+            update_fields.update(set(cf_mro))
+        for fieldname in cf_mro:
+            setattr(instance, fieldname, self._compute(instance, model, fieldname))
+        if update_fields:
+            return update_fields
+        return None
+
+    def precomputed(self, func):
+        """
+        Decorator for custom `save` methods, that expect local computed fields
+        to contain already updated values on enter.
+
+        By default local computed field values are only calculated once by the
+        `ComputedFieldModel.save` method after your own `save` method.
+
+        By placing this decorator on your save method, the values will be updated
+        before entering your method as well. Note that this comes for the price of
+        doubled local computed field calculations (before and after your save method).
```

Comment:
Why not skip computed field calculations in `ComputedFieldsModel.saver()` if `@precomputed` is used ?

Reply:
Well I currently strive for high data integrity, and we cannot guarantee this anymore, if someone does nasty late field modifications in `save` without another calculation.
Maybe this decorator should get a flag to suppress the second calculation? Something like `@precomputed(skip_second_calculation=True)`?

Reply:
Yes a flag would definitely work

Reply:
Implemented with latest commit as `@precomputed(skip_after=True)`.